contracts/meta-staking/src/state.rs

```diff
@@ -48,7 +59,197 @@ impl ConsumerInfo {
+    pub rewards_per_token: Decimal,
+}
+
+impl Default for ValidatorRewards {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ValidatorRewards {
+    pub fn new() -> Self {
+        ValidatorRewards {
+            rewards_per_token: Decimal::zero(),
+        }
+    }
+
+    pub fn calc_rewards(
+        &mut self,
+        rewards: Uint128,
+        total_tokens: Uint128,
+    ) -> Result<(), ContractError> {
+        let rewards_dec = Decimal::checked_from_ratio(rewards, total_tokens)?;
+
+        self.rewards_per_token = rewards_dec.checked_add(self.rewards_per_token)?;
+        Ok(())
+    }
 }
 
-/// Map<(consumer address, validator address), Validator>
+/// Map<(consumer address, validator address), Amount>
 pub const VALIDATORS_BY_CONSUMER: Map<(&Addr, &str), Uint128> = Map::new("validators_by_consumer");
+/// Map<(validator address, consumer address), Amount>
+pub const CONSUMERS_BY_VALIDATOR: Map<(&str, &Addr), Uint128> = Map::new("consumers_by_validators");
+
+#[cfg(test)]
+mod tests {
+    use cosmwasm_std::{Decimal, Uint128};
+
+    use super::{ConsumerInfo, ConsumerRewards, ValidatorRewards};
+    use crate::ContractError;
+
+    struct Staked {
+        total: Uint128,
+        consumer_one: Uint128,
+        consumer_two: Uint128,
+    }
+
+    impl Staked {
+        fn new() -> Self {
+            Staked {
+                total: Uint128::from(200_u128),
+                consumer_one: Uint128::from(100_u128),
+                consumer_two: Uint128::from(100_u128),
+            }
+        }
+
+        fn add_stake_to_consumer_one(&mut self, amount: Uint128) {
+            self.total += amount;
+            self.consumer_one += amount;
+        }
+
+        fn add_stake_to_consumer_two(&mut self, amount: Uint128) {
+            self.total += amount;
+            self.consumer_two += amount;
+        }
+
+        fn reduce_stake_to_consumer_one(&mut self, amount: Uint128) {
+            self.total -= amount;
+            self.consumer_one -= amount;
+        }
+
+        fn reduce_stake_to_consumer_two(&mut self, amount: Uint128) {
+            self.total -= amount;
+            self.consumer_two -= amount;
+        }
+    }
+
+    fn add_validator_rewards(
+        mut validator_rewards: ValidatorRewards,
+        rewards: Uint128,
+        total_staked: Uint128,
+    ) -> Result<ValidatorRewards, ContractError> {
+        validator_rewards.calc_rewards(rewards, total_staked)?;
+        Ok(validator_rewards)
+    }
+
+    fn calc_consumer_rewards(
+        validator_rewards: &ValidatorRewards,
+        consumer_rewards: ConsumerRewards,
+        staked: Uint128,
+    ) -> Result<ConsumerRewards, ContractError> {
+        let mut consumer_info = ConsumerInfo::new(Uint128::zero());
+
+        consumer_info.rewards = consumer_rewards;
+
+        consumer_info.calc_pending_rewards(validator_rewards.rewards_per_token, staked)?;
+        Ok(consumer_info.rewards)
+    }
+
+    #[test]
+    fn calculate_rewards() {
```

Comment:
I fixed all the issues mentioned @ethanfrey.

Here you can see the very basic calculation unit test i made, it only checks the math basically.
The test is on meta-staking, but the calculation is the same for both staking and provider.

Reply:
Tests look like the pass, and I guess I need to review meta-staking in general.

My big question here is I see is there is no reference to block height anywhere, and adjusting to it.
The dividing over block height before summing was the key issue I highlighted in the mesh-provider.
The fact there is no argument for height shows to me that we have different logic here than in mesh-provider.

You have convinced me the logic is general correct In meta-staking. I would like to see the similar test in mesh-provider

Reply:
Oh, I just saw https://github.com/CosmWasm/mesh-security/pull/40/commits/6c1a71beef858005d99b3b7295c1f0810b9b931b and you changed the logic to ignore heights.

Okay. That should be good enough for now. We can revisit the algorithm for something more precise / more efficient, but at least it seems relatively correct.